### ISSUE TITLE
Variable declare in taichi scope

### ISSUE THREAD

**Issue by IDCDC** [taichi-dev/taichi#5269]
Like `a = np.zeros((4,4))` in Python scope, I want to use `ti.field` to declare some variable in  Taichi scope, like in ti.func or ti.kernel. Unfortunately, I didn't find any code about this case. but I find that `ti.Matrix`  worked.
The next code can work, but the contents of the variable must be filled in, it goes against my purpose.
```
import taichi as ti

ti.init()

@ti.func
def a():
    x = ti.Matrix([[1, 2], [2, 1]])
    return x

@ti.kernel
def com():
    for _ in range(4):
        A = a()

com()
```
I've been tring this.

```
import taichi as ti

ti.init()

@ti.func
def a():
    x = ti.field(dtype=ti.f32, shape=(2,2))
    return x

@ti.kernel
def com():
    for _ in range(4):
        A = a()

com()
```
 and the error message is 
```
runfile('C:/Users/IDC/.spyder-py3/temp.py', wdir='C:/Users/IDC/.spyder-py3')
[Taichi] Starting on arch=x64
Traceback (most recent call last):

  File ~\.spyder-py3\temp.py:15 in <module>
    com()

  File G:\Anaconda\lib\site-packages\taichi\lang\kernel_impl.py:887 in wrapped
    raise type(e)('\n' + str(e)) from None

TaichiCompilationError: 
On line 13 of file "C:\Users\IDC\.spyder-py3\temp.py", in com:
        A = a()
            ^^^
On line 7 of file "C:\Users\IDC\.spyder-py3\temp.py", in a:
    x = ti.field(dtype=ti.f32, shape=(2,2))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Traceback (most recent call last):
  File "G:\Anaconda\lib\site-packages\taichi\lang\ast\ast_transformer_utils.py", line 24, in __call__
    return method(ctx, node)
  File "G:\Anaconda\lib\site-packages\taichi\lang\ast\ast_transformer.py", line 442, in build_Call
    node.ptr = func(*args, **keywords)
  File "G:\Anaconda\lib\site-packages\taichi\lang\util.py", line 295, in wrapped
    assert in_python_scope(), \
AssertionError: field cannot be called in Taichi-scope
```

**Comment by bobcao3** [comment]
You should declare fields outside of Taichi kernels, and then you can access them from within Taichi kernels.

**Comment by IDCDC** [comment]
Sorry, I didn't specify my requirements. I need to return a new field in Taichi scope, not only in `ti.kernel`, but also in `ti.func`. 
The first piece of code have to ensure every element is to be filled, but I just want to declare a field of zeros. Like `a = np.zeros(4)`, if declare a matrix, the shape is (100,100),  like the first piece of code, it's a lot of code to write. So I need a cleaner form.

**Comment by IDCDC** [comment]
> You should declare fields outside of Taichi kernels, and then you can access them from within Taichi kernels.

Does it means I can't declare a new variable in taichi scope by a cleaner way, like in `ti.kernel` and `ti.func`?

**Comment by neozhaoliang** [comment]
You can try this:

```python
from taichi.math import mat2

@ti.func
def foo():
     return mat2(0)
```

**Comment by IDCDC** [comment]
> @IDCDC For small matrices you can try this:
> 
> ```python
> from taichi.math import mat2
> 
> @ti.func
> def foo():
>     return mat2(0)
> ```

Thanks for your reply, it really helps. Where I can find more message? In https://docs.taichi.graphics/docs/field#declaration-2, I can only find fields, and besides, small matrices is also not my choice, I need big one.

**Comment by neozhaoliang** [comment]
@IDCDC It's not well documented yet, the source code is here:

https://github.com/taichi-dev/taichi/blob/master/python/taichi/math/mathimpl.py#L71-L114

Only mat2, mat3, mat4 are supplied and you can initialize them in various ways.

How large is your matrix? May be we can add an api for quick user defined matrices.

**Comment by IDCDC** [comment]
> @IDCDC It's not well documented yet, the source code is here:
> 
> https://github.com/taichi-dev/taichi/blob/master/python/taichi/math/mathimpl.py#L71-L114
> 
> Only mat2, mat3, mat4 are supplied and you can initialize them in various ways.
> 
> How large is your matrix? May be we can add an api for quick user defined matrices.

Yes, I found this source code already, thanks for your reply. It seems the biggest shape of vector and Matirx is four. I need to store a big array in FEM. For example: 
+  like 20 points of Gauss-Legendre quadrature points and corresponding weights.
+  If I choose a fine mesh, like $1024 \times 1024$, and store every cell error, or the coordinate of mesh node. 
Could I ask why we can't declare a large array in taichi scope as we do in the Python scope?  Does it for the compute efficiency?

**Comment by tengzhang48** [comment]
I am also new to Taichi and interested in FEM implementation. If you would like to define some matrices to be used in the FEM calculation, which are not a field variables of vertices and cells, why not just use numpy matrix? 

For the coordinate of the mesh, you should be able to use ti.vector.field. ti.matrix.field 3 by 3 should be enough for most FEM calculations as many variables are only 3 by 3 tensors although you can have many tensor variables. There may be higher rank tensors, but they probably can be expressed in a few low-rank tensors as well if needed. You can find from this example https://github.com/taichi-dev/taichi/blob/master/python/taichi/examples/simulation/fem128.py

For Gaussian integration points, my understanding is that you have a universal one for all the elements so you do not need to define them for each element.

**Comment by PENGUINLIONG** [comment]
> I want to take it as a class. And for different FEM method, if it can declare a big array will be necessary. Could I ask why we can't declare a large array in taichi scope as we do in the Python scope? Does it for the compute efficiency?

@IDCDC  Allocations in Taichi scope are actually not *memory allocations* but *register allocations* and they belongs to the concrete thread/invocation executing the kernel code. You usually can can allocate ~64 registers at maximal parallelism on modern graphics devices. So if you allocate a large number of registers in Taichi scope, I would expect a great loss of performance, because either the number of threads can be executed simultaneously is cut, or your computation intermediates fallback to the slower GDDR memory.

TL;DR, If your have large matricies to process I would recommend you to write dedicated kernels. It shouldn't cost many lines of code.

**Comment by IDCDC** [comment]
@tengzhang48 Thank you for the suggestion and example,  your question is some of the problems and attempts I have encountered in the past sevral months.

I achieved FEM method by numpy, but the time is too long, and it can be sloved by multiple processes. Because the basis is not the original basis, like linear basis, I choose multiscale basis.  Actually, it has the other parallel library in python, like multiprocessing. I also tried, but if I take the basis or the FEM space as a class, it can't pickle some data.

The reason why I want to declare any shape filed is if I use multiscale finite element method, I need define the FEM numerical solution as a `ti.func`, and used in a `ti.kernel`. Your advice take it as a numpy variable, it seems not work.

About Gaussian integration points, I also want to define it as a class.

**Comment by PENGUINLIONG** [comment]
> @PENGUINLIONG Thanks for your explanation, and help me to understand further. Does the dedicated kernel means depend on taichi?

Didn't quite understand the question, sry. I mean you can write a `@taichi.kernel` to do the matrix multiplication, or other large matrix operations, if that's what you wanted to do.

**Comment by tengzhang48** [comment]
My suggestion was to define global parameters with numpy matrix and still perform more computations in Taichi. The FEM code was made by the Taichi team and I am also leaning it now.

The process you mentioned seems to refinement of FEM by increasing the order of the shape functions. I am not sure whether "multiscale finite element" is the right description, although my understanding may be wrong. I feel it may be better to write a conventional FEM code in the Taichi framework first and then add more advanced techniques.

**Comment by neozhaoliang** [comment]
@IDCDC Currently all matrix operations are unrolled into the compiled kernel, that's why we only support small matrices.

If you want to process a big matrix, you can try load it into a field, or use `ti.ndarray`.

**Comment by IDCDC** [comment]
@tengzhang48 Thanks for your suggestion. If you are interested, the original "multiscale finite element" is from here, https://www.sciencedirect.com/science/article/pii/S0021999197956825

**Comment by IDCDC** [comment]
> Didn't quite understand the question, sry. I mean you can write a `@taichi.kernel` to do the matrix multiplication, or other large matrix operations, if that's what you wanted to do.
@PENGUINLIONG Thanks for the adivice. Well, actually, I write a `@ti.kernel` to do the matrix multiplication, but in this process, I also need to return a big matrix. That's why I asked.

**Comment by PENGUINLIONG** [comment]
> > Didn't quite understand the question, sry. I mean you can write a `@taichi.kernel` to do the matrix multiplication, or other large matrix operations, if that's what you wanted to do.
> 
> @PENGUINLIONG Thanks for the adivice. Well, actually, I write a `@ti.kernel` to do the matrix multiplication, but in this process, I also need to return a big matrix. That's why I asked. 
> 
> 

Ah I see. You can write to a field or ndarray passed into the Taichi scope. Return values are not necessarily 'returned'

**Comment by IDCDC** [comment]
> Ah I see. You can write to a field or ndarray passed into the Taichi scope. Return values are not necessarily 'returned'
@PENGUINLIONG  That's good idea, I have been think about this before I asked. If I choose this manner, I will declare a lot of variable. Considering the unnecessary memory overhead by this way, so I asked can I declare a variabla in `ti.kernel` and `ti.func` to avoid there are many variables in memory at the same time.

**Comment by IDCDC** [comment]
> It is currently impossible to do field allocations in runtime. Like we mentioned above, the variables in the kernels are actually registers not memory. If you need a large enough matrix that requires it to be put onto the memory, only thing close to it would be declaring fields ahead of time or declaring NDArrays as you use them.
> 
> In order to get a sense of scale here, how many of these matrices you would need for one iteration, and how large is each of them? It would be better for us to understand if you give some numbers, thanks! @IDCDC

Yes, from other people's answers in this discussion, I have understood the variables in taichi are actually registers. 

About the matrices number, as I mentioned, I will use $1024 \times 1024 \times 2$ for mesh node information or $400 \times 2$ for 2D Gaussian points and weights. 

I will try write a field and ndrray passed into taichi scope in the next step. 

Thanks to all those who answered this simple question, your detailed answers were very helpful.

**Comment by tengzhang48** [comment]
@IDCDC Thanks for sharing the paper. It is a very interesting work. I think it is still good to implement a conventional FEM first. The example is based on linear element without using Gaussian integration and higher order shape functions. After that you probably can have a better understanding of the data structure and code design to help you achieve the multiscale FEM.

**Comment by IDCDC** [comment]
> @IDCDC Thanks for sharing the paper. It is a very interesting work. I think it is still good to implement a conventional FEM first. The example is based on linear element without using Gaussian integration and higher order shape functions. After that you probably can have a better understanding of the data structure and code design to help you achieve the multiscale FEM.

Thanks for the advice, that's what I going to do next.
